### DISCUSSION TITLE
SQLite flexible typing

### DISCUSSION THREAD

**Discussion by capymind** [sqlalchemy/sqlalchemy#10741] [Usage Questions] [answered]
Hi, all.
SQLite supports so-called "flexible typing".

Using SQLAlchemy, it is possible.

```python

import sqlalchemy as sa

engine = sa.create_engine("sqlite://")
conn = engine.connect()

# create table.
stmt = """
CREATE TABLE user (
    attribute TEXT,
    value,
)       
"""
conn.execute(sa.text(stmt))
conn.commit()

# insert data into table.
stmt = sa.text("INSERT INTO user(attribute, value) VALUES(:attribute, :value)")
conn.execute(stmt.bindparams(attribute="stars", value=7777)) # Integer
conn.execute(stmt.bindparams(attribute="repo", value="SQLAlchemy"))  # String
conn.execute(stmt.bindparams(attribute="bytes", value=b'awesome')) # Byte
conn.commit()

# query data flexibly
data = conn.execute(sa.text("SELECT * FROM user")).all()
print(data)
conn.close()

# result
[('stars', 7777), ('repo', 'SQLAlchemy'), ('bytes', b'awesome')]
```

I want to define table using SQLAlchemy Core as follows:

```python
user_table = sa.Table(
    "user",
    metadata,
    sa.Column("attribute", sa.Text),
    sa.Column("value", "___I DON'T KNOW___"),
)
```

It's okay to set type to `NullType` or just skip for querying data.
However, if I want to use that table for creation or migration (using alembic), got CompileError such as
```python
    raise exc.CompileError(
sqlalchemy.exc.CompileError: (in table 'user', column 'value'): Can't generate DDL for NullType(); did you forget to specify a type on this Column?
```

I don't know how to harmonize this contradiction... in SQLite Dialect is there `ANYTHING` type?
I work around this if I define table for select only. However for migration and creating how to do that..

```python
user_table = sa.Table(
    "user",
    metadata,
    sa.Column("attribute", sa.Text),
    sa.Column("value"),
)

data = conn.execute(sa.select(user_table)).all()
print(data)

# result
[('stars', 7777), ('repo', 'SQLAlchemy'), ('bytes', b'awesome')]

# metadata.create_all(bind=engine) # -> error
# alembic --autogenerate revision   # -> error
```

relevant linkts
- SQLite Flexible Typing: https://www.sqlite.org/flextypegood.html
- SQLAlchemy NullType: https://docs.sqlalchemy.org/en/20/core/type_api.html#sqlalchemy.types.NullType

Thanks for reading this question!!

**Comment by zzzeek** [comment, marked as answer]
[make your own type](https://docs.sqlalchemy.org/en/20/core/custom_types.html#creating-new-types)

```py
import sqlalchemy.types as types

class ANYTHING(types.UserDefinedType):
    cache_ok = True

    def get_col_spec(self, **kw):
        return "ANYTHING"

```

**Reply by capymind** [reply]
Oh my... perfect.. it's like a magic.
I really appreciate it. Thank you for your reply!!
I hope I want to be a kind of alchemist someday...

**Reply by capymind** [reply]
By the way,
whenever I run `alembic revision --autogenerate` I got following notification.

```text
(.venv) rodi@LAPTOP-IINLT2LE:~/projects/wels$ alembic revision --autogenerate
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.autogenerate.compare] Detected type change from NUMERIC() to ANYTHING() on 'player_data.value'
INFO  [alembic.autogenerate.compare] Detected type change from NUMERIC() to ANYTHING() on 'starmap_data.value'
  Generating /home/rodi/projects/wels/wels/migrations/versions/b32fdd0fc61a_.py ...  done
```

I don't understand why `NUMERIC()` appears (I never assigned `NUMERIC` to `value` column..
Anyway I run `alembic upgrade head` and then I rerun `alembic revision --autogenerate` without modification, got same script again and again.

Sorry for brothering you. Where am i wrong...
Thanks.

```python
player_data = Table(
    "player_data",
    metadata,
    Column("player_id", Integer, index=True),
    Column("key", String, index=True),
    Column("value", ANYTHING),
    PrimaryKeyConstraint(
        "player_id",
        "key",
        name="player_data_pk",
    ),
)
```

**Reply by zzzeek** [reply]
type affinity is resolved by SQLAlchemy using the code below:

https://github.com/sqlalchemy/sqlalchemy/blob/896cac12737a2626e5a5115f406f9f2dc7509b90/lib/sqlalchemy/dialects/sqlite/base.py#L2297-L2349

this is based on SQLite's [specification](https://www.sqlite.org/datatype3.html#determination_of_column_affinity):

> For tables not declared as [STRICT](https://www.sqlite.org/stricttables.html), the affinity of a column is determined by the declared type of the column, according to the following rules in the order shown:
> 
>  If the declared type contains the string "INT" then it is assigned INTEGER affinity.
> 
>  If the declared type of the column contains any of the strings "CHAR", "CLOB", or "TEXT" then that column has TEXT affinity. Notice that the type VARCHAR contains the string "CHAR" and is thus assigned TEXT affinity.
> 
>  If the declared type for a column contains the string "BLOB" or if no type is specified then the column has affinity BLOB.
> 
>  If the declared type for a column contains any of the strings "REAL", "FLOA", or "DOUB" then the column has REAL affinity.
> 
>  Otherwise, the affinity is NUMERIC.


you will likely want to turn off type comparison for alembic using [compare_type=False](https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.environment.EnvironmentContext.configure.params.compare_type)

**Reply by capymind** [reply]
It takes some time for me to understand some concepts.. :)
Thanks a lot. It's really helpful for me!
